pkg/model/application.go

```diff
@@ -81,3 +81,13 @@ func (a *Application) IsOutOfSync() bool {
 func IsApplicationConfigFile(filename string) bool {
 	return filename == DefaultApplicationConfigFilename || strings.HasSuffix(filename, applicationConfigFileExtention)
 }
+
+func ToApplicationKind(kind string) (ApplicationKind, bool) {
```

Comment:
How about `ApplicationKindFromConfigKind`?
And I think this function should be better in the config package.
The reason is `config` package can import and refer things from the model package but not vice versa.

Reply:
I think a function should not know where will it be revoked, so explicitly mentioning `FromConfigKind` to show that this function will be revoked in `config` package does not look right to me. Besides, I keep the previous `cfg.Kind.ToApplicationKind`(code [here](https://github.com/pipe-cd/pipe/pull/2934/files#diff-a3d824da3c42420cd5cbb0a4a2c0e7b5bfddd819652788a0596d195dc6e31fa5R72)) as is in the `config` package and just move the implementation of it to this model package function so that this function can be used both from `config` package and other places (API as in this PR). I think that's good enough, how do you think about that?

Reply:
>  explicitly mentioning FromConfigKind to show that this function will be revoked in config package

No, I didn't mean that.
Instead, I tried to explicitly express what is being passed to that function with the `string` type.

The problem that makes my concern about putting the function in the model package is that `kind string` argument.
That kind could be `KubernetesApp`, `TerraformApp`, `ECSApp`, right?
It is the concept of `config` and should be stayed in the `config` package instead of being used in the `model` package since the `model` package is at a lower level compared to the `config` package in the dependency graph.

Reply:
I feel you, that's why I used the `kind string` primitive type instead of kind from `config` package. 👍 
But if you look closer, I just use the value name as `kind`, it can be whatever string, and I want to cast it to `ApplicationKind` which is a model's type. If an invalid value is passed, a false value will be returned. The name of argument as `kind` just because of the given value is going to be cast to ApplicationKind, that's all, no `config` or something related is mentioned here. 

Besides, what used the `cfg.Kind.ToApplicationKind()` (the config package code used places) will use that `config` package ToApplicationKind() function as is, nothing're changed.

In case we don't adopt my suggestion, we have to:
- Cast the primitive string value `matcher.Kind` (which be given from piped when it loads application configuration postSync and push to the controlplane in deployment chain trigger flow) to `config.Kind` which basically is an alias for string type.
- use defined `cfg.Kind.ToApplicationKind()` in `config` package to cast it to `ApplicationKind` - a model's type which is the type of value we want to filter in the datastore.

I think it's an unnecessary double-casting flow 🤔

Reply:
Well, I can say that, finally I can get what you're worried about 😄 
```golang
k := strings.TrimSuffix(upper, "APP")
```
the implementation contains this part which implied that values of the string passed here may contain `App` string that a `config` package concept, right?

In that case, lets do double-casting flow as I mentioned above, that's fine to me.

Reply:
> I just use the value name as kind, it can be whatever string, and I want to cast it to ApplicationKind which is a model's type. A simple cast to model's type from primitive type should be in the model package - I think. If an invalid value is passed, a false value will be returned. The name of argument as kind just because of the given value is going to be cast to ApplicationKind, that's all, no config or something related is mentioned here.

I don't think that changing the argument type to `string` could completely cut its relation to the `config` package.
Since that function is already designed to be dedicated for that config kind from internal with `k := strings.TrimSuffix(upper, "APP")`.

Reply:
Still I'm in the middle of getting what you want to do, I think ToApplicationKind should be in the config package because the logic highly depends on config specification (model shouldn't know that spec. It's just referred everywhere).

Reply:
> Since that function is already designed to be dedicated for that config kind from internal with k := strings.TrimSuffix(upper, "APP").

We're on the same page already 👍

Reply:
Ok. But wait. I have just realized that why do we need to convert that kind in RPC implementation?
Shouldn't they use `KUBERNETES`, `TERRAFORM`...?

Reply:
yeah, why don't you use `model.ApplicationKind` for `CreateDeploymentChainRequest_ApplicationMatcher`?

Reply:
To be clear, let me show an example

It should be

``` yaml
apiVersion: pipecd.dev/v1beta1
kind: KubernetesApp
spec:
  input:
    manifests:
      - deployment.yaml
      - service.yaml
    kubectlVersion: 1.18.0
    namespace: khanhtc-dev
  postSync:
    chain:
      applications:
        - name: simple-b
          kind: TERRAFORM
        - name: simple-c
```

not

``` yaml
apiVersion: pipecd.dev/v1beta1
kind: KubernetesApp
spec:
  input:
    manifests:
      - deployment.yaml
      - service.yaml
    kubectlVersion: 1.18.0
    namespace: khanhtc-dev
  postSync:
    chain:
      applications:
        - name: simple-b
          kind: TerraformApp
        - name: simple-c
```

Reply:
I see, I can say that the current `cfg.Kind.ToApplicationKind()` implementation allows both cases value passed above.

Reply:
KubernetesApp, TerraformApp is just a kind of configuration file not a kind of an application.
Besides KuberentesApp, ... we also have Piped, ControlPlane...

Reply:
> allows both cases value passed above.

Whichever it is, it's fine because it will eventually take the form of TERRAFORM.

Reply:
> yeah, why don't you use model.ApplicationKind for CreateDeploymentChainRequest_ApplicationMatcher?

@nakabonne I guess, we have to cast it in the upstream (piped before making a call to create deployment chain) or in the downstream (the gRPC implementation). I think it's better to keep the upstream more flexible on going to call, controlplane has more information to determine or validate. But I don't have a strong opinion on this, If you guys like the upstream validating, I will go that way, it's lgtm either.

Reply:
Let's do add `model.ToApplicationKind` which takes `Kubernetes, Terraform, etc` as input and returns `model.ApplicationKind`, then use it to cast the string we found in the application configuration postSync @nghialv @nakabonne

Reply:
> I think it's better to keep the upstream more flexible on going to call, controlplane has more information to determine or validate.

I think we should use the model.ApplicationKind for that matcher request. 
https://github.com/pipe-cd/pipe/blob/master/pkg/app/api/service/pipedservice/service.proto#L496

We can have `enum` validation (`[(validate.rules).enum.defined_only = true];`) on that field and that will be validated before being handled by the RPC.

Reply:
hold on

Reply:
Agree, I'm going that way now 🙆‍♂️

Reply:
>We can have enum validation ([(validate.rules).enum.defined_only = true];) on that field and that will be validated before being handled by the RPC.

Here we go.

Reply:
I just realized one point, if this `matcher.Kind` is not set, the default value (of type model.ApplicationKind) will be used; thus the kind will be `Kubernetes` by default. Is that okay with you guys? Since in some cases, application with the name mentioned as `matcher.Name` value does exist by not of kind `Kubernetes` and users who want to use that must mention both `matcher.Name` and `matcher.Kind` to enable control-plane to determine that application on deployment chain building process. @nghialv @nakabonne

Reply:
Nice catch! I got that point. (Maybe that was the reason why `string` was adopted.) 😄 

So I am fine to use the string for that field. Just want to add some patches as below:
- Add a comment to the proto to show why string was used but not model.Kind type
- In the RPC implementation, directly convert the string to model.Kind instead of using the function from the config package

Reply:
Sure 👍 Let's go that way. I find out that we can use proto3 optional here to make the `matcher.Kind` be a point to ApplicationKind by that require to update the dependencies of PipeCD project, lets thing about that later 👍

Reply:
For sure.

>In the RPC implementation, directly convert the string to model.Kind instead of using the function from the config package

Like it.

Reply:
@nghialv @nakabonne thanks for you guys' comments. Addressed by https://github.com/pipe-cd/pipe/pull/2934/commits/00044955ba8d036e616e6b171e5092c65f30ea82

Reply:
>  I find out that we can use proto3 optional here

Nice. Previously when I took a look at it, it was still under experimental status.

Reply:
https://github.com/protocolbuffers/protobuf/releases/tag/v3.15.0

From this release, it's marked as default but let address it by separated PR 😄